[JSC] Use WeakHashSet for the DeferredWorkTimer::Ticket's stored in JSGlobalObject

hyjorc1 opened the pull request:
#### 2aa59a0d48ddd9684584ccbbb6db8207d1a62733
<pre>
[JSC] Use WeakHashSet for the DeferredWorkTimer::Ticket&apos;s stored in JSGlobalObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=275518">https://bugs.webkit.org/show_bug.cgi?id=275518</a>
<a href="https://rdar.apple.com/129879654">rdar://129879654</a>

Reviewed by Yusuke Suzuki.

Use WeakHashSet for managing the TicketData&apos;s stored in JSGlobalObject
to avoid the manual unregistration in the destruction of TicketData.

* Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
(JSC::DeferredWorkTimer::TicketData::TicketData):
(JSC::DeferredWorkTimer::TicketData::create):
(JSC::DeferredWorkTimer::TicketData::cancel):
(JSC::DeferredWorkTimer::TicketData::globalObject):
(JSC::DeferredWorkTimer::doWork):
(JSC::DeferredWorkTimer::addPendingWork):
(JSC::DeferredWorkTimer::cancelPendingWorkSafe):
(JSC::DeferredWorkTimer::TicketData::~TicketData): Deleted.
(JSC::DeferredWorkTimer::TicketData::clearGlobalObject): Deleted.
* Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::~JSGlobalObject):
(JSC::JSGlobalObject::visitChildrenImpl):
(JSC::JSGlobalObject::addTicket):
(JSC::JSGlobalObject::clearTickets):
(JSC::JSGlobalObject::addObjectsForTicket): Deleted.
(JSC::JSGlobalObject::removeObjectsForTicket): Deleted.
(JSC::JSGlobalObject::clearObjectsForTicket): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/WaiterListManager.cpp:
(JSC::WaiterListManager::notifyWaiterImpl):
(JSC::WaiterListManager::unregister):
(JSC::Waiter::dump const):

Canonical link: <a href="https://commits.webkit.org/280180@main">https://commits.webkit.org/280180@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a1e9b0b431354a377f0031064f4d368b3a329b7a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55888 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35212 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8356 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58884 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6320 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42833 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6516 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45000 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4360 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57917 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33104 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48183 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26136 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29888 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5511 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4463 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/48964 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51857 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5782 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60473 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/55124 "Built successfully and passed tests") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5904 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52430 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48252 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51939 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12396 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/76885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31041 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/76885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32125 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33207 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31873 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -3288,28 +3290,15 @@ void JSGlobalObject::setWrapperMap(std::unique_ptr<WrapperMap>&& map)
-    addResult.iterator->value.at(i).set(vm(), this, scriptExecutionOwner);
-}
-void JSGlobalObject::removeObjectsForTicket(DeferredWorkTimer::Ticket ticket)
+void JSGlobalObject::addTicket(DeferredWorkTimer::Ticket ticket)
```
You need to take a lock

hyjorc1 replied:
Fixed.

Constellation commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
```diff
@@ -235,10 +230,10 @@ bool DeferredWorkTimer::cancelPendingWork(Ticket ticket)
-            cancelPendingWork(ticket);
-        m_tasks.append(std::make_tuple(ticket, [](DeferredWorkTimer::Ticket) mutable { }));
+    for (auto& ticket : *globalObject->m_tickets) {
+        if (!ticket.isCancelled())
```
You should keep ticket as Ref<TicketData> in the iteration body. This is weak ptr and can be nulled

hyjorc1 replied:
Fixed

Constellation commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
```diff
@@ -60,20 +60,15 @@ inline VM& DeferredWorkTimer::TicketData::vm()
+
+inline JSGlobalObject* DeferredWorkTimer::TicketData::globalObject()
+{
+    return target()->globalObject();
```
This does not work when target is also dead.

hyjorc1 replied:
Yes, encountered that crash. Fixed.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -316,7 +315,7 @@ void WaiterListManager::unregister(JSGlobalObject* globalObject)
             if (waiter->isAsync()) {
                 auto ticket = waiter->ticket(listLocker);
-                if (ticket->globalObject == globalObject) {
+                if (ticket->globalObject() == globalObject) {
```
This does not work when the target is also dead at this point.

hyjorc1 replied:
Yes, encountered that crash. Fixed.

Constellation commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
```diff
@@ -181,7 +172,7 @@ DeferredWorkTimer::Ticket DeferredWorkTimer::addPendingWork(VM& vm, JSObject* ta
     dependencies.append(Weak<JSCell>(target));
 
-    auto ticketData = makeUnique<TicketData>(globalObject, scriptExecutionOwner, WTFMove(dependencies));
+    RefPtr<TicketData> ticketData = TicketData::create(globalObject, scriptExecutionOwner, WTFMove(dependencies));
```
Let's take it as `auto`.

Constellation commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
```diff
@@ -42,16 +42,16 @@ static constexpr bool verbose = false;
 }
 
-inline DeferredWorkTimer::TicketData::~TicketData()
+inline RefPtr<DeferredWorkTimer::TicketData> DeferredWorkTimer::TicketData::create(JSGlobalObject* globalObject, JSObject* scriptExecutionOwner, Vector<Weak<JSCell>>&& dependencies)
```
Let's return `Ref<DeferredWorkTimer::TicketData>`.

Constellation commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
```diff
@@ -42,16 +42,16 @@ static constexpr bool verbose = false;
+inline RefPtr<DeferredWorkTimer::TicketData> DeferredWorkTimer::TicketData::create(JSGlobalObject* globalObject, JSObject* scriptExecutionOwner, Vector<Weak<JSCell>>&& dependencies)
 {
-    clearGlobalObject();
+    return adoptRef(new TicketData(globalObject, scriptExecutionOwner, WTFMove(dependencies)));
```
Let's return `adoptRef(*new ...);`

Constellation commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
```diff
@@ -122,7 +113,7 @@ void DeferredWorkTimer::doWork(VM& vm)
         // Remove ticket from m_pendingTickets since we are going to run it.
         // But we want to keep ticketData while running task since its globalObject ensures dependencies are strongly held.
-        std::unique_ptr<TicketData> ticketData = m_pendingTickets.take(pendingTicket);
+        RefPtr<TicketData> ticketData = m_pendingTickets.take(pendingTicket);
```
Let's take it as `auto`.

Constellation commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
```diff
@@ -96,13 +102,36 @@ class DeferredWorkTimer final : public JSRunLoopTimer {
     bool m_currentlyRunningTask { false };
     Deque<std::tuple<Ticket, Task>> m_tasks WTF_GUARDED_BY_LOCK(m_taskLock);
-    HashSet<std::unique_ptr<TicketData>> m_pendingTickets;
+    HashSet<RefPtr<TicketData>> m_pendingTickets;
```
Let's use `HashSet<Ref<TicketData>>`.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -3288,28 +3292,17 @@ void JSGlobalObject::setWrapperMap(std::unique_ptr<WrapperMap>&& map)
-{
-    Locker locker { cellLock() };
-    m_objectsForTicket->remove(ticket);
+    if (!m_weakTickets)
```
Let's emit `storeStoreFence()` before setting this to m_weakTickets.
```
if (!m_weakTickets) {
    auto weakTickets = makeUnique<WeakHashSet<DeferredWorkTimer::TicketData>>();
    WTF::storeStoreFence();
    m_weakTickets = WTFMove(weakTickets);
}
```

hyjorc1 replied:
Nice, I think this fence can ensure the initialization completeness that can mitigate the reordering effects even we have a lock here.

Constellation reviewed (approved):
r=me

hyjorc1 committed:
[JSC] Use WeakHashSet for the DeferredWorkTimer::Ticket's stored in JSGlobalObject
https://bugs.webkit.org/show_bug.cgi?id=275518
rdar://129879654

Reviewed by Yusuke Suzuki.

Use WeakHashSet for managing the TicketData's stored in JSGlobalObject
to avoid the manual unregistration in the destruction of TicketData.

* Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
(JSC::DeferredWorkTimer::TicketData::TicketData):
(JSC::DeferredWorkTimer::TicketData::create):
(JSC::DeferredWorkTimer::TicketData::cancel):
(JSC::DeferredWorkTimer::TicketData::globalObject):
(JSC::DeferredWorkTimer::doWork):
(JSC::DeferredWorkTimer::addPendingWork):
(JSC::DeferredWorkTimer::cancelPendingWorkSafe):
(JSC::DeferredWorkTimer::TicketData::~TicketData): Deleted.
(JSC::DeferredWorkTimer::TicketData::clearGlobalObject): Deleted.
* Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::~JSGlobalObject):
(JSC::JSGlobalObject::visitChildrenImpl):
(JSC::JSGlobalObject::addTicket):
(JSC::JSGlobalObject::clearTickets):
(JSC::JSGlobalObject::addObjectsForTicket): Deleted.
(JSC::JSGlobalObject::removeObjectsForTicket): Deleted.
(JSC::JSGlobalObject::clearObjectsForTicket): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/WaiterListManager.cpp:
(JSC::WaiterListManager::notifyWaiterImpl):
(JSC::WaiterListManager::unregister):
(JSC::Waiter::dump const):

Canonical link: https://commits.webkit.org/280180@main
